File: modules/core/src/encoder/sequence/SegmentSequence.ts
import {SequenceVersionMap} from './SequenceVersionMap';
import {TCModel} from '../../';
import {EncodingOptions} from '../EncodingOptions';
import {
  Segment,
  Fields,
} from '../../model';

export class SegmentSequence implements SequenceVersionMap {

  public '1': Segment[] = [
    Segment.CORE,
  ]

  public '2': Segment[] = [
    Segment.CORE,
  ]

  public constructor(tcModel: TCModel, options?: EncodingOptions) {

    if (tcModel.version === 2) {

      if (tcModel.isServiceSpecific) {

        /**
         * If it's service specific only, then the publisher TC String can be
         * stored in the cookie and would be transmitted if it's not for
         * storage.  So it's included regardless of whether or not it's for
         * saving or the cmp api to surface.
         */

        this['2'].push(Segment.PUBLISHER_TC);

      } else {

        const isForSaving = !!(options && !options.isForVendors);

        /**
         * including vendors disclosed only if it is for saving (to the global
         * scope) or supportOOB is turned on (either or both).  The compliment
         * of this being not for saving (surfaced to CMP) and no support of
         * OOB.
         */
        if (isForSaving || tcModel[Fields.supportOOB] === true) {

          this['2'].push(Segment.VENDORS_DISCLOSED);

        }

        if (!isForSaving) {

          /**
           * If a publisher does support OOB and they have narrowed the allowed
           * vendors to utilize it, then we should include the vendors allowed
           * segment.  If it is empty then there are no restrictions, if that
           * is intended to mean no support for OOB, then the flag should be
           * set for that instead.
           *
           */
          if (tcModel[Fields.supportOOB] && tcModel[Fields.vendorsAllowed].size > 0) {

            this['2'].push(Segment.VENDORS_ALLOWED);

          }

          /**
           * Always include the publisher TC segment as long as this TC string
           * is not intended to be saved in the global scope.
           */
          this['2'].push(Segment.PUBLISHER_TC);

        }

      }

    }

  }

}

File: GDPR-Phase2-Java/src/main/java/com/gdpr/phase2/encoder/sequence/SegmentSequence.java
package com.gdpr.phase2.encoder.sequence;

import com.gdpr.phase2.model.Segments;
import com.gdpr.phase2.Utils.TCModel;

import java.util.ArrayList;
import java.util.List;

public class SegmentSequence extends VersionMap {
    private SegmentSequence() {
    }
    private static final SegmentSequence instance = new SegmentSequence();
    public static SegmentSequence getInstance() {
        return instance;
    }

    public final List<String> one = new ArrayList<String>(){{
        add(Segments.core);
    }};
    public final List<String> two = new ArrayList<String>(){{
        add(Segments.core);
    }};

    public SegmentSequence(TCModel tcModel,Boolean isForSaving) {
        if(tcModel.getServiceSpecific()) {
            /**
             * If it's service specific only, then the publisher TC String can be
             * stored in the cookie and would be transmitted if it's not for
             * storage
             */
            if(isForSaving) {
                this.two.add(Segments.publisherTC);
            }
        } else {
            /**
             * this is a globally scoped string.
             *
             * If the publisher supports OOB, then the disclosed vendors vector will
             * be added to both the transmission string and the storage string
             */
            if(tcModel.getSupportOOB() && tcModel.vendorsDisclosed.getSet().size()>0) {
                this.two.add(Segments.vendorsDisclosed);
            }

            /**
             * If this string is not for saving then the vendors allowed vector can
             * be added – Otherwise, it should be omitted because this is a publisher
             * specific setting and we don't want to store publisher-specific
             * settings in the global cookie.
             */
            if(!isForSaving) {
                if(tcModel.vendorsAllowed.getSet().size()>0) {
                    this.two.add(Segments.vendorsAllowed);
                }
                if(tcModel.publisherConsents.getSet().size() > 0 ||
                tcModel.publisherLegitimateInterest.getSet().size() > 0 ||
                tcModel.getNumCustomPurposes() > 0 ||
                tcModel.publisherCustomConsents.getSet().size() > 0 ||
                tcModel.publisherCustomLegitimateInterest.getSet().size() > 0) {
                    this.two.add(Segments.publisherTC);
                }
            }
        }
    }
}
